link_test.go

```diff
@@ -664,3 +667,151 @@ func TestLinkStats64UnmarshalBinary(t *testing.T) {
 		})
 	}
 }
+
+func TestLinkXDP(t *testing.T) {
+	// BPF compilation requires a high RLIMIT_MEMLOCK
+	n := uint64(1024 * 1024 * 10)
+	err := unix.Setrlimit(unix.RLIMIT_MEMLOCK, &unix.Rlimit{Cur: n, Max: n})
+	if err != nil {
+		t.Fatalf("failed to increase RLIMIT_MEMLOCK: %s", err)
+	}
+
+	// establish a netlink connections
+	conn, err := Dial(nil)
+	if err != nil {
+		t.Fatalf("failed to establish netlink socket")
+	}
+	defer conn.Close()
+
+	// construct dummy interface to test XDP program against
+	if err := conn.Link.New(&LinkMessage{
+		Family: unix.AF_UNSPEC,
+		Index:  1001,
+		Flags:  unix.IFF_UP,
+		Attributes: &LinkAttributes{
+			Name: "dummy0",
+			Info: &LinkInfo{Kind: "dummy"},
+		},
+	}); err != nil {
+		t.Fatalf("failed to construct dummy interface: %s", err)
+	}
+
+	// get info for the dummy interface
+	interf, err := conn.Link.Get(1001)
+	if err != nil {
+		conn.Link.Delete(interf.Index)
+		t.Fatalf("could not get dummy interface info: %s", err)
+	}
+
+	// compile a BPF test program. If it fails error out of the tests
+	// and clean up dummy interface.
+	prog, err := ebpf.NewProgram(&ebpf.ProgramSpec{
+		Type: ebpf.XDP,
+		Instructions: asm.Instructions{
+			asm.LoadImm(asm.R0, int64(2), asm.DWord),
+			asm.Return(),
+		},
+		License: "MIT",
+	})
+	if err != nil {
+		conn.Link.Delete(interf.Index)
+		t.Fatalf("failed to compile test bpf program: %s", err)
+	}
+
+	prog2, err := ebpf.NewProgram(&ebpf.ProgramSpec{
+		Type: ebpf.XDP,
+		Instructions: asm.Instructions{
+			asm.LoadImm(asm.R0, int64(4), asm.DWord),
+			asm.Return(),
+		},
+		License: "MIT",
+	})
+	if err != nil {
+		conn.Link.Delete(interf.Index)
+		t.Fatalf("failed to compile test bpf program: %s", err)
+	}
+	// Use the file descriptor of the program
+	progFD := int32(prog.FD())
+	progFD2 := int32(prog2.FD())
+
+	message := LinkMessage{
+		Family:     unix.AF_UNSPEC,
+		Index:      interf.Index,
+		Attributes: &LinkAttributes{},
+	}
+
+	tests := []struct {
```

Comment:
Do these test cases rely on order in some way to progress through different kernel states? If so, the setup/teardown logic should be factored out so that each test can run independently.

Reply:
The goal here is to test loading a program, attempt to replace it with a wrong `expected_fd` (which should fail), replace it with a correct `expected_fd` and as a last step clear the XDP program from the link.

The tests could be split up in "load program", "clear program" and "replace program".

Reply:
I think the 2nd replace test is failing due to order. If the test is an actual sequence, they should be sequenced.

Especially if (for whatever reason) we should decide to run tests in parallel at some point, this would fail bad.

Reply:
I refactored the tests so the stage are tested independently.

The tests seem to fail because the `exptected_fd` property (and the `REPLACE` flag is only introduced from kernel `5.7` and onwards, before it was not supported (hence the `invalid argument` error).

5.6: https://elixir.bootlin.com/linux/v5.6/source/net/core/dev.c#L8662
5.7: https://elixir.bootlin.com/linux/v5.7/source/net/core/dev.c#L8674

Reply:
I've added a `kernelMinReq` check to skip the test for kernels `<5.7`.

Reply:
Could  you add that as a comment to the test skip line, so consumers can understand why and what changed in the kernel versions?

Reply:
I've done my best to keep it as brief as possible :sweat_smile: 
https://github.com/jsimonetti/rtnetlink/pull/113/files#diff-a8f0967251c1f1c9f29b71e2b55cd0c3adcb7d62e9fa1826c566a46672e4ce7bR271